<a href="https://colab.research.google.com/github/mark-narusov/yandex_practicum_ds/blob/main/nlp_classification/nlp_narusov_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Предобработка-датафрейма" data-toc-modified-id="Предобработка-датафрейма-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка датафрейма</a></span></li><li><span><a href="#Предобработка-текста" data-toc-modified-id="Предобработка-текста-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Предобработка текста</a></span></li><li><span><a href="#Разбиение-на-выборки" data-toc-modified-id="Разбиение-на-выборки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Разбиение на выборки</a></span></li><li><span><a href="#Подбор-метода-цифровизации-текста" data-toc-modified-id="Подбор-метода-цифровизации-текста-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Подбор метода цифровизации текста</a></span><ul class="toc-item"><li><span><a href="#Мешок-слов" data-toc-modified-id="Мешок-слов-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Мешок слов</a></span></li><li><span><a href="#Биграммы" data-toc-modified-id="Биграммы-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Биграммы</a></span></li><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span><code>TF-IDF</code></a></span></li></ul></li><li><span><a href="#Подбор-модели-и-гиперпараметров" data-toc-modified-id="Подбор-модели-и-гиперпараметров-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Подбор модели и гиперпараметров</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Catboost" data-toc-modified-id="Catboost-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span><em>Catboost</em></a></span><ul class="toc-item"><li><span><a href="#text_features" data-toc-modified-id="text_features-7.2.1"><span class="toc-item-num">7.2.1&nbsp;&nbsp;</span><code>text_features</code></a></span></li><li><span><a href="#Мешок-слов" data-toc-modified-id="Мешок-слов-7.2.2"><span class="toc-item-num">7.2.2&nbsp;&nbsp;</span>Мешок слов</a></span></li><li><span><a href="#Подбор-learning_rate" data-toc-modified-id="Подбор-learning_rate-7.2.3"><span class="toc-item-num">7.2.3&nbsp;&nbsp;</span>Подбор <code>learning_rate</code></a></span></li></ul></li></ul></li><li><span><a href="#Подбор-порога" data-toc-modified-id="Подбор-порога-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Подбор порога</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию.

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75.

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели.
3. Сделайте выводы.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Введение

Итак, перед нами задача обучения с учителем, а именно — бинарной классификации. Будем преобразовывать текстовый признак-предиктор в понятный для наших алгоритмов язык матриц. Для классификации комментариев на токсичные и нетоксичные попробуем использовать как подход "мешок слов", так и расчёт взвешенной частоты употребления слова `TF-IDF`.  

Успех преобразования данных и подбора модели будем оценивать с помощью метрики качества *F1*, среднего гармонического между метриками *Recall* и *Precision*.

## Импорт библиотек

Импортируем нужные библиотеки:

In [ ]:
import pandas as pd
import numpy as np
import scipy

# для отображения progress bar при применении функции
from tqdm import tqdm
tqdm.pandas()

# для очистки текста от неинформативных символов
import re

# nltk — библиотека для работы с текстовыми признаками
import nltk
from nltk.tokenize import TreebankWordTokenizer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split, GridSearchCV

from sklearn.linear_model import LogisticRegression

# ключевая метрика для проекта
from sklearn.metrics import f1_score

from catboost import CatBoostClassifier

## Предобработка датафрейма

Загружаем данные:

In [ ]:
try:
    df = pd.read_csv('/datasets/toxic_comments.csv')
except:
    df = pd.read_csv('toxic_comments.csv')

Посмотрим на данные:

In [ ]:
df.head(5)

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

Итак, у нас всего один признак-предиктор, сам текст комментария. Связь между текстом и классом `toxic` нам и предстоит установить с помощью машинного обучения.

Так как перед нами задача классификации, посмотрим на баланс классов целевой переменной:

In [ ]:
df.toxic.value_counts(normalize=True).round(2)

0    0.9
1    0.1
Name: toxic, dtype: float64

Учтём, что в целевом признаке дисбаланс классов. Будем разбивать на выборки со стратификацией по целевому признаку, а также подберём порог классификации для максимизации *F1*.

С помощью `assert` установим, что
1. В датафрейме нет пропусков
2. В датафрейме нет дублированных наблюдений
3. Все `id` комментариев уникальны (количество уникальных `id` соответствует длине датафрейма)

In [ ]:
assert all([df.isna().sum().sum() == 0,
            df.duplicated().sum() == 0,
            df.iloc[:, 0].nunique() == df.shape[0]])

Порядок.  
Исправим дублированный индекс и сотрём его название:

In [ ]:
df = df.set_index(df.columns[0])
df.index.name = None

Также переименуем столбец `toxic` в `is_toxic` для соответствия булевому смыслу колонки и переведём её в тип `bool`.

In [ ]:
df = df.rename({'toxic': 'is_toxic'}, axis=1)
df['is_toxic'] = df['is_toxic'].astype('bool')
df.head(5)

text  is_toxic
0  Explanation\nWhy the edits made under my usern...     False
1  D'aww! He matches this background colour I'm s...     False
2  Hey man, I'm really not trying to edit war. It...     False
3  "\nMore\nI can't make any real suggestions on ...     False
4  You, sir, are my hero. Any chance you remember...     False

Проверим, что текстовый признак уже в нужном нам формате:

In [ ]:
assert all(df['text'] == df['text'].values.astype('U'))

In [ ]:
df.shape

(159292, 2)

Итак, перед нами 159 292 текста, к каждому из которых присвоено булево значение.

## Предобработка текста

Чтобы токенизировать текст нам предстоит:  
1. Очистить текст от лишних символов.
2. Привести слова к нижнему регистру и убрать неинформативные стоп-слова.
3. Лемматизировать оставшиеся слова, перед этим присвоив им теги с частью речи — это нужно для корректности работы лемматизатора WordNetLemmatizer.

сохраняем лемматизатор в соответствующую переменную:

In [ ]:
lemmatizer = WordNetLemmatizer()

Для быстрой работы функции сохраним используемое регулярное выражение в переменную с помощью `re.compile()`, а также сохраним в переменную множество (`set`) стоп-слов — поиск по значению в этом классе занимает константное время O(1).

In [ ]:
# это выражение ищет все символы, не являющиеся буквами
regex_compiled = re.compile(r'[^A-Za-z]')
# множество стоп-слов
eng_stopwords = set(stopwords.words('English'))

Теперь напишем хоть и не самую читаемую, но оптимизированную функцию для предобработки текста:

In [ ]:
def preprocess_text(text: str) -> str:
    '''
    Функция принимает на вход строку, возвращает её в лемматизированном
    виде.
    Порядок операций:
    1. Все символы кроме букв заменяются на пробел, слова разбиваются через пробел.
    2. Из списка удаляются стоп-слова, оставшиеся приводятся к нижнему регистру
    3. С помощью nltk.pos_tag() каждому токену присваиватся тег соответствующей
       части речи
    4. Лемматизатор WordNet затем лемматизирует токены в соответствие с тегом части речи
    5. На выходе токены склеиваются функцией .join(), с пробелом в качестве разделителя
    '''

    token_pos_tuples = nltk.pos_tag([
        token.lower() for token in re.sub(
            regex_compiled, ' ', text).split()
        if token not in eng_stopwords
    ])
    lemmas = [
        # существительные
        lemmatizer.lemmatize(token, pos='n') if pos.startswith('N')
        # глаголы
        else lemmatizer.lemmatize(token, pos='v') if pos.startswith('V')
        # прилагательные
        else lemmatizer.lemmatize(token, pos='a') if pos.startswith('J')
        # наречия
        else lemmatizer.lemmatize(token, pos='r') if pos.startswith('R')
        # всё остальное
        else lemmatizer.lemmatize(token)
        for token, pos in token_pos_tuples
    ]
    return ' '.join(lemmas)

Посмотрим на пример работы функции:

In [ ]:
print(f'''Текст до предобработки:
{df['text'][0]}
Длина текста:
{len(df['text'][0])}


Текст после предобработки:
{preprocess_text(df['text'][0])}
Длина текста:
{len(preprocess_text(df['text'][0]))}''')

Текст до предобработки: 
Explanation
Why the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27
Длина текста:
264


Текст после предобработки: 
explanation why edits make username hardcore metallica fan revert they vandalisms closure gas i vote new york doll fac and please remove template talk page since i retire
Длина текста:
170


Функция работает корректно. Применим её к столбцу `text` (с отображением прогресса), запишем результат в столбец `text_lemmatized` в том же датафрейме.

In [ ]:
%%time
df['text_lemmatized'] = df['text'].progress_apply(preprocess_text)

100%|██████████████████████████████████| 159292/159292 [06:21<00:00, 417.26it/s]

CPU times: user 6min 14s, sys: 6.84 s, total: 6min 20s
Wall time: 6min 21s


In [ ]:
df.head(5)

text  is_toxic  \
0  Explanation\nWhy the edits made under my usern...     False   
1  D'aww! He matches this background colour I'm s...     False   
2  Hey man, I'm really not trying to edit war. It...     False   
3  "\nMore\nI can't make any real suggestions on ...     False   
4  You, sir, are my hero. Any chance you remember...     False   

                                     text_lemmatized  
0  explanation why edits make username hardcore m...  
1  d aww he match background colour i seemingly s...  
2  hey man i really try edit war it guy constantl...  
3  more i make real suggestion improvement i wond...  
4              you sir hero any chance remember page

Итак, теперь у нас есть корпус. Запишем его в переменную `full_corpus`:

In [ ]:
full_corpus = df['text_lemmatized'].values

Перед тем как преобразовывать корпус в мешки слов и биграммы, разобьём выборку на обучающую, тестовую и валидационную.

## Разбиение на выборки

При разбиении учтём сильный дисбаланс классов (9:1), сделаем так чтобы баланс классов в каждой из выборок был пропорционален балансу классов исходного датафрейма.

In [ ]:
def train_valid_test_split(df: pd.DataFrame, target: str) -> \
    tuple[pd.DataFrame, pd.Series,
          pd.DataFrame, pd.Series,
          pd.DataFrame, pd.Series]:
    '''
    Функция принимает датафрейм df и строку target со столбцом-таргетом,
    Отдаёт три пары датафрейм-серия для трёх выборок со стратифицированным
    целевым признаком.
    '''
    X = df.copy()
    y = X.pop(target)

    X_train, X_test_valid, y_train, y_test_valid = \
    train_test_split(X, y, test_size=0.4, random_state=42, stratify=y)

    X_test, X_valid, y_test, y_valid = \
    train_test_split(X_test_valid,
                     y_test_valid,
                     test_size=0.5,
                     random_state=42,
                     stratify=y_test_valid)
    # Проверяем, что выборки действительно разбиты в отношении 3:1:1
    # и классы стратифицированны
    assert all([
        round(X_train.shape[0] / y_test.shape[0], 2) == 3,
        round(X_valid.shape[0] / y_test.shape[0], 2) == 1,
        np.sum(y_test) - np.sum(y_valid) <= 1,
        np.sum(y_train) - np.sum(y_valid) * 3 <= 3
    ])

    return X_train, y_train, X_valid, y_valid, X_test, y_test

In [ ]:
X_train, y_train, X_valid, y_valid, X_test, y_test = \
train_valid_test_split(df, 'is_toxic')

In [ ]:
train_corpus = X_train['text_lemmatized'].values
valid_corpus = X_valid['text_lemmatized'].values
test_corpus = X_test['text_lemmatized'].values

## Подбор метода цифровизации текста

В нашем проекте мы будем использовать три этапа подбора на валидационной выборке перед тем как тестировать модель на тестовой выборке:
1. Подбор метода цифровизации текста
2. Подбор модели и её гиперпараметров
3. Подбор порога классификации для выбранной модели

Начнём с подбора метода цифровизации текста. Для этого импортируем логистическую регрессию, с помощью которой будем измерять качество преобразования.

In [ ]:
lr = LogisticRegression(max_iter=1000, random_state=42)

### Мешок слов

Создадим счётчик слов и "обучим" его на тренировочной выборке:

In [ ]:
count_vect = CountVectorizer()

In [ ]:
count_vect.fit(train_corpus)

CountVectorizer()

Создадим два мешка слов для выборок:

In [ ]:
bow_train = count_vect.transform(train_corpus)
bow_valid = count_vect.transform(valid_corpus)

Обучим модель на тренировочной выборке:

In [ ]:
lr.fit(bow_train, y_train)

LogisticRegression(max_iter=1000, random_state=42)

Посчитаем *F1-score* для валидационной выборки:

In [ ]:
f1_score(y_valid, lr.predict(bow_valid)).round(3)

0.762

Уже неплохо. Попробуем другие методы.

### Биграммы

Проделаем ту же процедуру для биграмм:

In [ ]:
count_vect_bigram = CountVectorizer(ngram_range=(2, 2))

In [ ]:
count_vect_bigram.fit(train_corpus)

CountVectorizer(ngram_range=(2, 2))

In [ ]:
bigram_train = count_vect_bigram.transform(train_corpus)
bigram_valid = count_vect_bigram.transform(valid_corpus)

In [ ]:
lr.fit(bigram_train, y_train)

LogisticRegression(max_iter=1000, random_state=42)

In [ ]:
f1_score(y_valid, lr.predict(bigram_valid)).round(3)

0.411

Ужасный скор, попробуем расчёт `TF-IDF`.

### `TF-IDF`

In [ ]:
count_tf_idf = TfidfVectorizer()

Подберём лучшие параметры расчёта относительных частот слов `TF-IDF` с помощью класса `sklearn.pipeline.Pipeline`, где результат всё той же логистической регрессии по метрике качестве *F1* будет оценочной функцией.

In [ ]:
pipeline = Pipeline([
    ('tf_idf', count_tf_idf),
    ('lr', lr),
])
parameters = {
    'tf_idf__max_df': (0.25, 0.5, 0.75),
    'tf_idf__ngram_range': [(1, 1), (2, 2), (1, 2)]
}

grid_search_tf_idf = \
GridSearchCV(pipeline, parameters, cv=3, scoring='f1', verbose=3)

In [ ]:
grid_search_tf_idf.fit(train_corpus, y_train)
pass

Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV 1/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 1);, score=0.710 total time=   6.0s
[CV 2/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 1);, score=0.715 total time=   5.3s
[CV 3/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 1);, score=0.719 total time=   5.6s
[CV 1/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(2, 2);, score=0.182 total time=  25.3s
[CV 2/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(2, 2);, score=0.179 total time=  23.7s
[CV 3/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(2, 2);, score=0.185 total time=  23.8s
[CV 1/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 2);, score=0.674 total time=  35.8s
[CV 2/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 2);, score=0.674 total time=  35.8s
[CV 3/3] END tf_idf__max_df=0.25, tf_idf__ngram_range=(1, 2);, score=0.683 total time=  35.2s
[CV 1/3] END tf_idf__max_df=0.5, tf_idf__ngram_range=(1, 1);, score=0.710 total time=   7.0s
[

In [ ]:
print(f'''Лучший скор F1 на кросс-валидации: \
{grid_search_tf_idf.best_score_:.3f}

Гиперпараметры лучшего расчёта TF-IDF:
{grid_search_tf_idf.best_estimator_.steps[0][1]}
TfidfVectorizer(ngram_range=(1, 1))''')

Лучший скор F1 на кросс-валидации: 0.714

Гиперпараметры лучшего расчёта TF-IDF: 
TfidfVectorizer(max_df=0.25)
TfidfVectorizer(ngram_range=(1, 1))


Получилось лучше чем с биграммами, но хуже чем с мешком слов.  

**Вывод:**  
Итак, лучшим методом преобразования оказался самый простой — мешок слов. Заранее создадим мешок слов для тестовой выборки и приступим к подбору модели и её гиперпараметров.

In [ ]:
bow_test = count_vect.transform(test_corpus)

## Подбор модели и гиперпараметров

С помощью инструментала библиотеки `scipy` для работы с разреженными матрицами объединим тренировочные и валидационные выборки для подбора параметров логистической регрессии при помощи кросс-валидации:

In [ ]:
bow_full_train = scipy.sparse.vstack((bow_train, bow_valid))
y_full_train = pd.concat([y_train, y_valid])

### Логистическая регрессия

Импортируем логистическую регрессию и сразу зафиксируем достаточно большое кол-во итерации для сходимости:

In [ ]:
lr = LogisticRegression(max_iter=1000, random_state=42)

Зададим сетку возможных параметров для обучения:

In [ ]:
lr_param_grid = {
    'class_weight': [None, 'balanced'],
    'solver': ['lbfgs', 'newton-cg'],
    'C': [1,2,6]
}

Создадим инстанцию класса `GridSearchCV`. Будем подбирать гиперпараметры на трёх фолдах и оценивать модель с помощью целевой метрики — *F1*.

In [ ]:
lr_grid_search_cv = GridSearchCV(lr,
                                 lr_param_grid,
                                 cv=3,
                                 verbose=3,
                                 scoring='f1')

In [ ]:
lr_grid_search_cv.fit(bow_full_train, y_full_train)

Fitting 3 folds for each of 12 candidates, totalling 36 fits
[CV 1/3] END C=1, class_weight=None, solver=lbfgs;, score=0.765 total time=  13.0s
[CV 2/3] END C=1, class_weight=None, solver=lbfgs;, score=0.757 total time=  10.1s
[CV 3/3] END C=1, class_weight=None, solver=lbfgs;, score=0.765 total time=  17.1s
[CV 1/3] END C=1, class_weight=None, solver=newton-cg;, score=0.765 total time=  20.0s
[CV 2/3] END C=1, class_weight=None, solver=newton-cg;, score=0.757 total time=  17.6s
[CV 3/3] END C=1, class_weight=None, solver=newton-cg;, score=0.765 total time=  25.3s
[CV 1/3] END C=1, class_weight=balanced, solver=lbfgs;, score=0.763 total time=  12.2s
[CV 2/3] END C=1, class_weight=balanced, solver=lbfgs;, score=0.755 total time=  11.0s
[CV 3/3] END C=1, class_weight=balanced, solver=lbfgs;, score=0.773 total time=  13.2s
[CV 1/3] END C=1, class_weight=balanced, solver=newton-cg;, score=0.763 total time=  41.9s
[CV 2/3] END C=1, class_weight=balanced, solver=newton-cg;, score=0.755 total

GridSearchCV(cv=3, estimator=LogisticRegression(max_iter=1000, random_state=42),
             param_grid={'C': [1, 2, 6], 'class_weight': [None, 'balanced'],
                         'solver': ['lbfgs', 'newton-cg']},
             scoring='f1', verbose=3)

In [ ]:
print(f'''Лучший скор F1 на кросс-валидации: \
{lr_grid_search_cv.best_score_:.3f}

Гиперпараметры лучшей модели: {lr_grid_search_cv.best_params_}''')

Лучший скор F1 на кросс-валидации: 0.769

Гиперпараметры лучшей модели: {'C': 2, 'class_weight': None, 'solver': 'lbfgs'}


Так как подбирать модель *CatBoost*'а мы будем не с помощью кросс-валидации, а на отложенной валидационной выборке, то и сравнивать логистическую регрессию стоит по скоре на отложенной выборке.

Запишем модель с лучшими гиперпаматреами в переменную `lr_best` и обучим её на тренировочной выборке:

In [ ]:
lr_best = LogisticRegression(**lr_grid_search_cv.best_params_,
                             max_iter=1000,
                             random_state=42)

In [ ]:
lr_best.fit(bow_train, y_train)

LogisticRegression(C=2, max_iter=1000, random_state=42)

Посчитаем скор на валидационной выборке:

In [ ]:
f1_score(y_valid, lr_best.predict(bow_valid)).round(3)

0.766

### *Catboost*

Попробуем обучить *Catboost* в двух вариантах:
1. Просто передадим ему лемматизированный текст как `text_feature`
2. Обучим его на мешке слов

#### `text_features`

Приведём признаки в понятный для алгоритма формат:

In [ ]:
train_df = pd.DataFrame({'text_feature': train_corpus})
valid_df = pd.DataFrame({'text_feature': valid_corpus})

Будем оценивать *Catboost* метрикой *F1* и останавливать обучение если метрика не улучшается уже 200 итераций:

In [ ]:
cb_text = CatBoostClassifier(text_features=['text_feature'],
                             eval_metric='F1',
                             early_stopping_rounds=200,
                             verbose=100,
                             random_state=42)

Обучаем *CatBoost* на лемматизированном тексте:

In [ ]:
cb_text.fit(train_df, y_train, eval_set=(valid_df, y_valid))
pass

Learning rate set to 0.097736
0:	learn: 0.7039864	test: 0.7343701	best: 0.7343701 (0)	total: 275ms	remaining: 4m 34s
100:	learn: 0.7545187	test: 0.7581584	best: 0.7581584 (100)	total: 23.5s	remaining: 3m 29s
200:	learn: 0.7734352	test: 0.7673970	best: 0.7680675 (167)	total: 47.4s	remaining: 3m 8s
300:	learn: 0.7868302	test: 0.7693653	best: 0.7693653 (300)	total: 1m 15s	remaining: 2m 54s
400:	learn: 0.7969291	test: 0.7705432	best: 0.7714286 (380)	total: 1m 40s	remaining: 2m 30s
500:	learn: 0.8041463	test: 0.7701928	best: 0.7717958 (453)	total: 2m 4s	remaining: 2m 4s
600:	learn: 0.8088853	test: 0.7721914	best: 0.7721914 (599)	total: 2m 28s	remaining: 1m 38s
700:	learn: 0.8139508	test: 0.7716808	best: 0.7725381 (656)	total: 2m 52s	remaining: 1m 13s
800:	learn: 0.8177125	test: 0.7713051	best: 0.7725381 (656)	total: 3m 14s	remaining: 48.3s
Stopped by overfitting detector  (200 iterations wait)

bestTest = 0.7725381415
bestIteration = 656

Shrink model to first 657 iterations.


Лучший скор на валидационной выборке:

In [ ]:
round(cb_text.best_score_['validation']['F1'], 4)

0.7725

Пока что лучшая модель.

#### Мешок слов

Теперь попробуем обучить *CatBoost* на разреженной матрице частот (мешке слов):

In [ ]:
cb_bow = CatBoostClassifier(eval_metric='F1', verbose=100, random_state=42)

In [ ]:
cb_bow.fit(bow_train, y_train, eval_set=(bow_valid, y_valid))
pass

Learning rate set to 0.097736
0:	learn: 0.4134330	test: 0.4132431	best: 0.4132431 (0)	total: 480ms	remaining: 7m 59s
100:	learn: 0.6405415	test: 0.6386691	best: 0.6386691 (100)	total: 50.9s	remaining: 7m 33s
200:	learn: 0.7060327	test: 0.6947005	best: 0.6947005 (200)	total: 1m 39s	remaining: 6m 35s
300:	learn: 0.7317897	test: 0.7135603	best: 0.7135603 (300)	total: 2m 32s	remaining: 5m 53s
400:	learn: 0.7491676	test: 0.7237282	best: 0.7237282 (398)	total: 3m 21s	remaining: 5m 1s
500:	learn: 0.7631322	test: 0.7332965	best: 0.7332965 (497)	total: 4m 6s	remaining: 4m 5s
600:	learn: 0.7696704	test: 0.7353372	best: 0.7359148 (564)	total: 4m 58s	remaining: 3m 17s
700:	learn: 0.7719320	test: 0.7358974	best: 0.7359148 (564)	total: 5m 45s	remaining: 2m 27s
800:	learn: 0.7753742	test: 0.7377469	best: 0.7377469 (779)	total: 6m 37s	remaining: 1m 38s
900:	learn: 0.7767143	test: 0.7381127	best: 0.7383434 (893)	total: 7m 30s	remaining: 49.5s
999:	learn: 0.7784902	test: 0.7384953	best: 0.7385346 (981)	

Лучший скор на валидационной выборке:

In [ ]:
round(cb_bow.best_score_['validation']['F1'], 3)

0.739

Алгоритм, работающий на лемматизированном тексте, показал себя лучше, чем тот, который работал с матрицей частот.

#### Подбор `learning_rate`

Попробуем улучшить результат, уменьшив `learning_rate` так, чтобы алгоритм сходился только к концу обучения, к тысячной итерации:

In [ ]:
cb_text_2 = CatBoostClassifier(text_features=['text_feature'],
                             eval_metric='F1',
                             learning_rate=0.03,
                             early_stopping_rounds=200,
                             verbose=100,
                             random_state=42)

In [ ]:
cb_text_2.fit(train_df, y_train, eval_set=(valid_df, y_valid))
pass

0:	learn: 0.7039864	test: 0.7343701	best: 0.7343701 (0)	total: 205ms	remaining: 3m 24s
100:	learn: 0.7201882	test: 0.7455121	best: 0.7458641 (98)	total: 22.4s	remaining: 3m 19s
200:	learn: 0.7377303	test: 0.7534704	best: 0.7534704 (200)	total: 48s	remaining: 3m 10s
300:	learn: 0.7504753	test: 0.7566901	best: 0.7571655 (265)	total: 1m 14s	remaining: 2m 53s
400:	learn: 0.7592068	test: 0.7596408	best: 0.7599577 (388)	total: 1m 42s	remaining: 2m 32s
500:	learn: 0.7665606	test: 0.7638791	best: 0.7638791 (500)	total: 2m 6s	remaining: 2m 6s
600:	learn: 0.7723986	test: 0.7655620	best: 0.7665322 (580)	total: 2m 33s	remaining: 1m 42s
700:	learn: 0.7770767	test: 0.7677227	best: 0.7679103 (691)	total: 2m 56s	remaining: 1m 15s
800:	learn: 0.7804031	test: 0.7700928	best: 0.7700928 (800)	total: 3m 19s	remaining: 49.5s
900:	learn: 0.7845886	test: 0.7712190	best: 0.7712190 (900)	total: 3m 43s	remaining: 24.5s
999:	learn: 0.7879805	test: 0.7721563	best: 0.7722910 (991)	total: 4m 8s	remaining: 0us

bestT

In [ ]:
round(cb_text_2.best_score_['validation']['F1'], 4)

0.7723

Получилось немного хуже, чем с `learning_rate` по умолчанию.

**Вывод:**

Градиентный бустинг библиотеки *Catboost*, обученный на лемматизированном тексте, показал себя как лучшая модель. Приступим к подбору порога.

## Подбор порога

Так как классы в целевом признаки сильно несбалансированны, порог предсказания по умолчанию величиной в 0.5 может отличаться от оптимального для максимизации *F1*.   

Может быть этот порог слишком низкий, и тогда уменьшается *Precision*, или слишком высокий, и тогда уменьшается *Recall*.  
Экспериментально подберём оптимальный порог на валидационной выборке.

Запишем скоры модели для положительного класса для каждого объекта в переменную `y_scores`:

In [ ]:
y_scores = cb_text.predict_proba(valid_df)[:, 1]

Всего будет 101 возможных порогов. Будем считать значение метрики *F1* на каждом из них и добавлять его в список `f1_scores`:

In [ ]:
thresholds = np.linspace(0, 1, num=101)
f1_scores = []

for threshold in thresholds:
    # y_pred — булев массив
    y_pred = y_scores >= threshold
    f1_scores.append(f1_score(y_valid, y_pred))

results_df = pd.DataFrame({'threshold': thresholds, 'f1': f1_scores})

Посмотрим на десять лучших порогов по значению метрики *F1*:

In [ ]:
results_df.set_index('threshold').sort_values(by='f1',
                                              ascending=False).head(10)

f1
threshold          
0.37       0.788045
0.36       0.787333
0.32       0.786947
0.34       0.786531
0.33       0.786301
0.35       0.786251
0.31       0.785391
0.30       0.784630
0.38       0.784333
0.39       0.783718

**Вывод:**

Как видим из результатов подбора, для максимизации *F1* порог по умолчанию был слишком высок — модель допускала слишком много ошибок первого рода и тем самым уменьшала *Precision* даже больше, чем увеличивала *Recall*.

Запишем лучший порог в переменную `thresholds`:

In [ ]:
threshold = results_df.sort_values(
    by='f1', ascending=False).head(1)['threshold'].values[0]
threshold

0.37

Объявим функцию `predict_with_threshold`, предсказывающую целевой признак с помощью переданных ей модели, данных, а также порога.

In [ ]:
def predict_with_threshold(model, X, threshold: float) -> np.array:
    '''
    Функция предсказыввает целевой признак бинарной классификации по
    переданному порогу threshold с помощью модели model на данных X.
    Возвращает булев массив.
    '''
    y_scores = model.predict_proba(X)[:, 1]
    return y_scores >= threshold

Можно переходить к тестированию модели.

## Тестирование модели

Запишем тестовый корпус в формат датафрейм:

In [ ]:
test_df = pd.DataFrame({'text_feature': test_corpus})

Итак, посчитаем итоговый *F1*-скор модели:

In [ ]:
round(f1_score(
    y_test, predict_with_threshold(cb_text, test_df, threshold)),
    3)

0.794

Заметно выше чем минимум, установленный заказчиком.

## Выводы

Итак, перед нами стояла задача выбора алгоритма бинарной классификации с текстовыми признаками для предсказания, является ли комментарий токсичным и стоит ли его отправить на модерацию.

Сперва мы добавили в датафрейм лемматизированную версию исходных текстов, очищенную от лишних символов.

Затем мы разбили датафрейм на тренировочную, валидационную и тестовую выборки в отношении 3:1:1.

После этого мы провели три этапа подбора параметров итогового алгоритма на валидационной выборке
1. Попробовали три разных метода цифровизации текста: мешок слов, биграммы, `TF-IDF`. Лучшим оказался мешок слов
2. Подобрали модель и её гиперпараметры: лучшей оказался градиентный бустинг *Catboost*, обученный на лемматизированных текстах как `text_features`. Мешок слов в итоге не понадобился.
3. Подобрали оптимальный порог классификации для оптимизации метрики *F1*: $0.37$

**В итоге наш алгоритм классификации набрал значение *F1*-меры величиной в $0.79$, что на четыре сотых больше, чем установленный заказчиком минимум в $0.75$**